node/src/main/scala/coop/rchain/node/grpc.scala

```diff
@@ -60,24 +60,22 @@ object GrpcServer {
   class ReplImpl(runtime: Runtime)(implicit scheduler: Scheduler) extends ReplGrpc.Repl {
     import RholangCLI.{buildNormalizedTerm, evaluate}
 
-    def exec(reader: Reader): Future[ReplResponse] = buildNormalizedTerm(reader) match {
-      case Left(er) =>
-        Future.successful(ReplResponse(s"Error: $er"))
-      case Right(term) =>
-        val evalAttempt: Either[Throwable, Unit] =
-          evaluate(runtime.reducer, term).attempt.unsafeRunSync
-
-        Task
-          .pure(
-            evalAttempt match {
+    def exec(reader: Reader): Future[ReplResponse] =
+      Task
+        .delay(buildNormalizedTerm(reader))
+        .flatMap {
+          case Left(er) =>
+            Task.now(s"Error: $er")
+          case Right(term) =>
+            evaluate(runtime.reducer, term).attempt.map {
               case Left(ex) => s"Caught boxed exception: $ex"
               case Right(_) =>
                 s"Storage Contents:\n ${StoragePrinter.prettyPrint(runtime.store)}"
             }
-          )
-          .map(ReplResponse(_))
-          .runAsync
-    }
+        }
+        .map(ReplResponse(_))
+        .executeAsync
+        .runAsync
```

Comment:
TBH I don't see the code that changes the scheduler the operation is running on. Which line exactly ?

Reply:
Please take a look at my comments in the Jira ticket. We have to ensure an asynchronous boundary with `.executeAsync`. Also now `buildNormalizedTerm` is wrapped in a `Task.delay`. And I joined back the two parts off the execution (evaluate and pretty print) with a `.map`. The result is one `Task` recipe which runs on the provided scheduler.

Reply:
OK, read it @sebtomba . It makes sense thanks. I am wondering, if you had removed the `executeAsync` it would still be executed on implicit scheduler (namely `grpc-default-executor`) but would be susceptible to be batched with other operations right? Would it still be trigerring error that Henry found?

Reply:
Follow up comment @sebtomba . Why this bothers me is that I thought that `Scheduler`s created with `Scheduler.io` helper should be unbounded in threads number and if thread is being used then another one is being created.

Reply:
Without `executeAsync` the `evaluate` part runs on `grpc-default-executor` and `prettyPrint` on `repl-io`. Actually, we need to investigate if this is a Monix bug or expected behaviour.
The maxreaders exceeded bug is actually not a bug but a configuration issue. LMDB assigns by default reader slots to threads for the whole application lifecycle. It doesn't matter to which thread pool this thread belongs. Henry is working on this. This PR solves a different problem.